### PR TITLE
Perform batch registration to Meeting and Webinar

### PR DESCRIPTION
**Pull request by maaaruf**
Hello, @Jericho, 
I would like to express my sincere gratitude to you for maintaining this repository, which has been an immense help to me.

I have added two new methods named `PerformBatchRegistration` to `Meeting.cs` and `Webinar.cs`, as well as added their corresponding signatures to `IWebinar.cs` and `IMeeting.cs`. In addition, I have also created two new models named `BatchRegistrant` and `BatchRegistrantInfo`, which are used to send and receive information about registrants during the batch registration API calling.

**Here is a brief overview of the changes:**

* Added `PerformBatchRegistration` methods to `Meeting.cs` and `Webinar.cs`, which enable batch registration of multiple participants at once.

*  Added the signatures for these methods to `IWebinar.cs` and `IMeeting.cs`.

* Created `BatchRegistrant` and `BatchRegistrantInfo` models to to send and receive information about registrants during the batch registration API calling.

* Added `BatchRegistrant`, `BatchRegistrantInfo` and `BatchRegistrant[]`, `BatchRegistrantInfo[]` to `ZoomNetJsonSerializerContext.cs`.

These changes should make it easier to register multiple participants at once, and provide a more streamlined registration process for meetings and webinars.

Please let me know if you have any questions, suggestions, or concerns about these changes.

### CODE REVIEW

File: Source/ZoomNet/Resources/Meetings.cs [hunk 1]
```diff
@@ -529,6 +529,43 @@ public Task<RegistrantInfo> AddRegistrantAsync(long meetingId, string email, str
...
+		/// <param name="registrants">An array of registrants.</param>
+		/// <param name="autoApprove">Indicates if the registrant should be automatically approved.</param>
+		/// <param name="registrantsConfirmationEmail">Indicates if send confirmation Email to Registrants.</param>
+		/// <param name="cancellationToken">The cancellation token.</param>
+		/// <returns>
+		/// An array of <see cref="BatchRegistrantInfo" />.
+		/// </returns>
+		public async Task<BatchRegistrantInfo[]> PerformBatchRegistrationAsync(long meetingId, BatchRegistrant[] registrants, bool autoApprove = false, bool registrantsConfirmationEmail = false, CancellationToken cancellationToken = default)
+		{
+			if (registrants.Any() == false || registrants.Count() > 30)
```
Comment (Jericho) [inline Source/ZoomNet/Resources/Meetings.cs:545]: You should make sure registrants is not null before checking the number of items in the enumeration.

Here's an example when I do something similar. I check that at least one item has been provided and no more than 5 (in your case the max value is 30, but otherwise the logic is the same): https://github.com/Jericho/ZoomNet/blob/develop/Source/ZoomNet/Resources/Chat.cs#L187-L188

File: Source/ZoomNet/Resources/Meetings.cs [hunk 2]
```diff
@@ -671,7 +708,7 @@ public Task<Poll[]> GetPollsAsync(long meetingId, CancellationToken cancellation
 			return _client
 				.GetAsync($"meetings/{meetingId}/polls")
 				.WithCancellationToken(cancellationToken)
-				.AsObject<Poll[]>("polls");
+				.AsObject<Poll[]>(propertyName: "polls");
```
Comment (Jericho) [inline Source/ZoomNet/Resources/Meetings.cs:711]: Why was this changed?

File: Source/ZoomNet/Resources/Webinars.cs [hunk 3]
```diff
@@ -553,6 +553,41 @@ public Task<RegistrantInfo> AddRegistrantAsync(long webinarId, string email, str
...
+		/// <param name="webinarId">The webinar ID.</param>
+		/// <param name="registrants">An array of registrants.</param>
+		/// <param name="autoApprove">Indicates if the registrant should be automatically approved.</param>
+		/// <param name="cancellationToken">The cancellation token.</param>
+		/// <returns>
+		/// An array of <see cref="BatchRegistrantInfo" />.
+		/// </returns>
+		public async Task<BatchRegistrantInfo[]> PerformBatchRegistrationAsync(long webinarId, BatchRegistrant[] registrants, bool autoApprove = false, CancellationToken cancellationToken = default)
+		{
+			if (registrants.Any() == false || registrants.Count() > 30)
```
Comment (Jericho) [inline Source/ZoomNet/Resources/Webinars.cs:568]: See my comment on `IMeetings.PerformBatchRegistrationAsync`

File: Source/ZoomNet/Resources/Webinars.cs [hunk 4]
```diff
@@ -553,6 +553,41 @@ public Task<RegistrantInfo> AddRegistrantAsync(long webinarId, string email, str
...
+			{
+				{ "auto_approve", autoApprove },
+				{ "registrants", registrants },
+			};
+
+			var response = await _client
+				.PostAsync($"webinars/{webinarId}/batch_registrants")
+				.WithJsonBody(data)
+				.WithCancellationToken(cancellationToken)
+				.AsRawJsonDocument()
```
Comment (Jericho) [inline Source/ZoomNet/Resources/Webinars.cs:583]: See my comment on `IMeetings.PerformBatchRegistrationAsync`

File: Source/ZoomNet/Models/BatchRegistrantInfo.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+		/// Gets or sets the URL for this registrant to join the meeting or webinar.
+		/// </summary>
+		[JsonPropertyName("join_url")]
+		public string JoinUrl { get; set; }
+
+		/// <summary>
+		/// Gets or sets she participant PIN code. Which is used to authenticate audio participants before they join the meeting.
+		/// </summary>
+		[JsonPropertyName("participant_pin_code")]
+		public string ParticipantPinCode { get; set; }
```
[the remark this answers is not in the document]
Comment (maaaruf) [reply Source/ZoomNet/Models/BatchRegistrantInfo.cs]: Yes, I fixed it.

File: Source/ZoomNet/Resources/IMeetings.cs [hunk 6]
```diff
@@ -261,6 +261,19 @@ public interface IMeetings
...
+		/// </summary>
+		/// <param name="meetingId">The meeting ID.</param>
+		/// <param name="registrants">An array of registrants.</param>
+		/// <param name="autoApprove">Indicates if the registrant should be automatically approved.</param>
+		/// <param name="registrantsConfirmationEmail">Indicates if send confirmation Email to Registrants.</param>
+		/// <param name="cancellationToken">The cancellation token.</param>
+		/// <returns>
+		/// An array of <see cref="BatchRegistrantInfo" />.
+		/// </returns>
+		Task<BatchRegistrantInfo[]> PerformBatchRegistrationAsync(long meetingId, BatchRegistrant[] registrants, bool autoApprove = false, bool registrantsConfirmationEmail = false, CancellationToken cancellationToken = default);
```
[the remark this answers is not in the document]
Comment (maaaruf) [reply Source/ZoomNet/Resources/IMeetings.cs]: Thank you for the suggestion. I've changed it accordingly

File: Source/ZoomNet/Resources/Meetings.cs (hunk 2 above)
Comment (maaaruf) [reply Source/ZoomNet/Resources/Meetings.cs]: It was done by mistake maybe.

File: Source/ZoomNet.IntegrationTests/Tests/Meetings.cs
Comment (Jericho) [inline Source/ZoomNet.IntegrationTests/Tests/Meetings.cs:1]: Excellent, thanks for adding this integration test

**Review by Jericho — APPROVED** [review]
Looks good

File: Source/ZoomNet/Json/ZoomNetJsonSerializerContext.cs [hunk 7]
```diff
@@ -18,7 +19,7 @@ namespace ZoomNet.Json
 	[JsonSerializable(typeof(ZoomNet.Models.AuthenticationSettings))]
 	[JsonSerializable(typeof(ZoomNet.Models.AuthenticationType))]
 	[JsonSerializable(typeof(ZoomNet.Models.BatchRegistrant))]
-	[JsonSerializable(typeof(ZoomNet.Models.BatchRegistrant[]))]
+	[JsonSerializable(typeof(IEnumerable<ZoomNet.Models.BatchRegistrant>))]
```
Comment (maaaruf) [inline Source/ZoomNet/Json/ZoomNetJsonSerializerContext.cs:22]: Hello @Jericho, I'm unable to find this line in both the main and development branches, which is causing an exception. Is there any way to resolve this issue? It seems that the same problem is also occurring with the released NuGet package.

image

Comment (Jericho) [reply Source/ZoomNet/Json/ZoomNetJsonSerializerContext.cs:22]: This is caused by the fact that registrants are not converted to an array prior to serialization in PerformBatchRegistrationAsync. This code:
```
{ "registrants", registrants },
```

should be:
```
{ "registrants", registrants.ToArray() },
```

I'll fix it.

Comment (maaaruf) [reply Source/ZoomNet/Json/ZoomNetJsonSerializerContext.cs:22]: Thanks a lot mate. It's working perfectly now.
